jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,571 @@
+				+ ", type = " + targetType);  //$NON-NLS-1$
+		/* Check the primitive type and MemoryAddress against the ValueLayout */
+		} else if (targetType != void.class) {
+			if (!ValueLayout.class.isInstance(targetLayout)) {
+				throw new IllegalArgumentException("ValueLayout is expected: layout = " + targetLayout); //$NON-NLS-1$
+			}
+			/* Check the size and kind of the ValueLayout for the primitive types and MemoryAddress */
+			validateValueLayoutSize(targetLayout, targetType);
+			validateValueLayoutKind(targetLayout, targetType);
+		}
+	}
+	
+	/* Check the size of the specified primitive layout to determine whether it matches the specified type */
+	private static void validateValueLayoutSize(MemoryLayout TypeLayout, Class<?> targetType) {
+		int layoutSize = (int)TypeLayout.bitSize();
+		boolean mismatchedSize = false;
+		
+		switch (layoutSize) {
+		case 8:
+			/* the 8-bits layout in Java only matches with byte in C */
+			if (targetType != byte.class) {
+				mismatchedSize = true;
+			}
+			break;
+		case 16:
+			/* The 16-bits layout is shared by char and short
+			 * given the char size is 16 bits in Java.
+			 */
+			if ((targetType != char.class) && (targetType != short.class) ) {
+				mismatchedSize = true;
+			}
+			break;
+		case 32:
+			/* The 32-bits layout is shared by boolean, int and float
+			 * given the boolean type is treated as int in Java.
+			 */
+			if ((targetType != boolean.class)
+			&& (targetType != int.class)
+			&& (targetType != float.class)
+			) {
+				mismatchedSize = true;
+			}
+			break;
+		case 64:
+			/* The 32-bits layout is shared by long, double and the MemoryAddress class
+			 * given the corresponding pointer size is 32 bits in C.
+			 */
+			if ((targetType != long.class)
+			&& (targetType != double.class)
+			&& (targetType != MemoryAddress.class)
+			) {
+				mismatchedSize = true;
+			}
+			break;
+		default:
+			mismatchedSize = true;
+			break;
+		}
+		
+		if (mismatchedSize) {
+			throw new IllegalArgumentException("Mismatched size between the layout and the type: layout = " //$NON-NLS-1$
+				+ TypeLayout + ", type = " + targetType.getSimpleName());  //$NON-NLS-1$
+		}
+	}
+	
+	/* Check the kind (type) of the specified primitive layout to determine whether it matches the specified type */
+	private static void validateValueLayoutKind(MemoryLayout targetLayout, Class<?> targetType) {
+		boolean mismatchType = false;
+		
+		if (!targetLayout.toString().contains("[abi/kind=")) { //$NON-NLS-1$
+			throw new IllegalArgumentException("The layout's ABI Class is undefined: layout = " + targetLayout); //$NON-NLS-1$
+		}
+		
+		/* Extract the kind from the specified layout with the ATTR_NAME "abi/kind".
+		 * e.g. b32[abi/kind=INT]
+		 */
+		TypeKind kind = (TypeKind)targetLayout.attribute(TypeKind.ATTR_NAME)
+				.orElseThrow(() -> new IllegalArgumentException("The layout's ABI Class is empty: layout = " + targetLayout)); //$NON-NLS-1$
```

Comment:
I don't think there's much value in appending the empty optional in the message; I suggest the message can be simplified to `"The layout's ABI class is empty"`.

Reply:
Updated the message.